### PR TITLE
Add button to generate partial URL

### PR DESCRIPTION
**Pull request by jkempster34**
### Trello card
https://trello.com/c/l8pD2kl7

### Context
The CRM team requested a way to generate partial URLs based on the event name and start date.

### Changes proposed in this pull request
- Add button to generate partial URL

### Guidance to review
- I originally did this on name/date change. But I thought that could be annoying if you choose a custom name, then change name or date.
- Tested this by testing the individual methods, do I need to set up the HTML template like some of the other controllers instead/as well?
- Should this have some feedback when name or date is not set? Would this just be an error message (suppose I could just hide the button until they are set)?
- I replace whitespace with hyphens to match the other events, are there any other characters that shouldn't appear in a URL?

### CODE REVIEW

File: app/webpacker/controllers/internal_events_controller.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { Controller } from 'stimulus';
+
+export default class extends Controller {
+  static targets = ['name', 'startAt', 'partialUrl'];
+
+  get nameValue() {
```
Comment (ethax-ross) [inline app/webpacker/controllers/internal_events_controller.js:6]: I wouldn't bother with these getters, its just a layer of indirection it doesn't add anything IMO

File: app/webpacker/controllers/internal_events_controller.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  get partialUrlValue() {
+    return this.partialUrlTarget.value;
+  }
+
+  set partialUrlValue(value) {
+    this.partialUrlTarget.value = value;
+  }
+
+  generateButtonClicked(event) {
```
Comment (ethax-ross) [inline app/webpacker/controllers/internal_events_controller.js:22]: The function should describe what its doing, that way when you see it in the HTML its easier to understand, for example:

```
click->internal-events#generateButtonClicked

vs

click->internal-events#populatePartialUrl
```

File: app/views/internal/events/new.html.erb [hunk 3]
```diff
@@ -27,14 +31,23 @@
...
     <%= f.datetime_field :end_at, class: class_names("datetime", "govuk-input",
                                                      { "govuk-input--error": @event.errors[:end_at].any? }) %>
   <% end %>
 
+  <%= f.govuk_text_field :readable_id,
+                         label: { text: 'Event partial URL' },
+                         hint: { text: 'You can generate the partial URL from the event name and start at date' },
+                         data: { target: "internal-events.partialUrl" } %>
+  <%= f.button "Generate partial URL",
+               class: "call-to-action-button button form-button",
```
Comment (peteryates) [inline app/views/internal/events/new.html.erb:48]: I think you should be ok with just `button` here:

```suggestion
               class: "button",
```

File: app/views/internal/events/new.html.erb [hunk 4]
```diff
@@ -59,5 +72,5 @@
       <% end %>
     <% end %>
   </div>
-  <%= f.button "Submit for review", class: "call-to-action-button button" %>
+  <%= f.button "Submit for review", class: "call-to-action-button button form-button" %>
```
Comment (peteryates) [inline app/views/internal/events/new.html.erb:75]: And here

```suggestion
  <%= f.button "Submit for review", class: "button" %>
```

File: app/webpacker/controllers/internal_events_controller.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    return `${date}-${name}`;
+  }
+
+  formatDate(dateTimeString) {
+    if (dateTimeString === '') return '';
+
+    const date = dateTimeString.substring(0, dateTimeString.indexOf('T'));
+    const [year, month, day] = date.split('-');
+    return year.slice(-2) + month + day;
```
Comment (ethax-ross) [inline app/webpacker/controllers/internal_events_controller.js:41]: This is pretty hard to follow - would it be cleaner to construct a new `Date` from the string and then output that in a format?

File: spec/javascript/controllers/internal_events_controller_spec.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import InternalEventsController from 'internal_events_controller.js';
+
+describe('InternalEventsController', () => {
+  let controller;
+
+  beforeEach(() => {
+    controller = new InternalEventsController();
```
[the remark this answers is not in the document]
Comment (ethax-ross) [reply spec/javascript/controllers/internal_events_controller_spec.js:7]: I would replace them or you'll end up doubling up; if you test the output of the controller actions by asserting against the HTML it will also let you refactor your methods without needing to update your tests, which is beneficial

File: app/webpacker/controllers/internal_events_controller.js (hunk 5 above)
Comment (peteryates) [inline app/webpacker/controllers/internal_events_controller.js:41]: Could this be simplified by using a date object? Something along the lines of:

```
const date = new Date(dateTimeString);
return date.toLocaleDateString();
```

Comment (peteryates) [reply app/webpacker/controllers/internal_events_controller.js:41]: Ignore that, assumed JavaScript had some kind of built-in date formatter 🤦🏽‍♂️

Comment (jkempster34) [reply app/webpacker/controllers/internal_events_controller.js:41]: Pete made the same comment, not sure if there is a way to do this.

Comment (peteryates) [reply app/webpacker/controllers/internal_events_controller.js:41]: Yeah it's a bit of a faff because JavaScript's date functionality is pretty limited. [toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString) has lots of options but it's a total faff compared to `#strftime`.

```suggestion
    const date = new Date(dateTimeString);
    return date.toLocaleString('en-GB', { '🤷🏽‍♂️' });
```
